src/Symfony/Component/PropertyAccess/PropertyAccessor.php

```diff
@@ -457,13 +458,79 @@ private function writeProperty(&$object, $property, $singular, $value)
 
             $object->$property = $value;
         } elseif (self::ACCESS_TYPE_MAGIC === $access[self::ACCESS_TYPE]) {
-            $object->{$access[self::ACCESS_NAME]}($value);
+            $this->callMethod($object, $access[self::ACCESS_NAME], $value);
         } else {
             throw new NoSuchPropertyException($access[self::ACCESS_NAME]);
         }
     }
 
     /**
+     * Throws a {@see UnexpectedTypeException} as in PHP 7 when using PHP 5.
+     *
+     * @param object $object
+     * @param string $method
+     * @param mixed  $value
+     *
+     * @throws UnexpectedTypeException
+     * @throws \Exception
+     */
+    private function callMethod($object, $method, $value) {
+        if (PHP_MAJOR_VERSION >= 7) {
+            try {
+                $object->{$method}($value);
+            } catch (\TypeError $e) {
+                throw $this->createUnexpectedTypeException($object, $method, $value);
+            }
+
+            return;
+        }
+
+        $that = $this;
+        set_error_handler(function ($errno, $errstr) use ($object, $method, $value, $that) {
+            if (E_RECOVERABLE_ERROR === $errno && false !== strpos($errstr, sprintf('passed to %s::%s() must', get_class($object), $method))) {
+                throw $that->createUnexpectedTypeException($object, $method, $value);
+            }
+
+            return false;
+        });
+
+        try {
+            $object->{$method}($value);
+            restore_error_handler();
+        } catch (\Exception $e) {
+            // Cannot use finally in 5.5 because of https://bugs.php.net/bug.php?id=67047
```

Comment:
in fact, since autoloading would not happen in the finally block, the bug would not apply if I'm not wrong.
But still, finally doesn't exist on PHP 5.3